util.go

```diff
@@ -29,3 +37,103 @@ func min(values ...uint32) uint32 {
 	}
 	return m
 }
+
+type segmentedBuffer struct {
+	cap     uint64
+	pending uint64
+	len     uint64
+	bm      sync.Mutex
+	b       [][]byte
+}
+
+// NewSegmentedBuffer allocates a ring buffer.
+func NewSegmentedBuffer(initialCapacity uint32) segmentedBuffer {
+	return segmentedBuffer{cap: uint64(initialCapacity), b: make([][]byte, 0)}
+}
+
+func (s *segmentedBuffer) Len() int {
+	return int(atomic.LoadUint64(&s.len))
```

Comment:
What was the motivation for the atomics? We end up doing a lot of atomic operations just to allow checking the length without taking a lock.

(I haven't profiled it so it may be fine, I'm just wondering).

Reply:
reads will be regularly checking if the length has extended. my intuition was this would be faster than locking, reading, and unlocking.

Reply:
Yeah, my real concern is whether that outweighs the cost of multiple atomic operations (and potential memory barriers) while we're holding a lock anyways. But it works so I have no real objections.

Reply:
i suspect a lot of that has to do with the list of buffers rather than the length atomics 🤷‍♂️ 

I'm also going to revert back to uint32's throughout, as they were before. There was an underflow that i was confirming, which is why i bumped them up to 64 bits.

Reply:
SGTM.